### ISSUE TITLE
Docs: distributed writes guide

### ISSUE THREAD

**Issue by wjones127** [lancedb/lance#3271]
Show how to do distributed writes in Python.

**Comment by ahaapple** [comment]
I was thinking about this question today. If I use S3 express as storage and want to implement distributed multiple nodes to build the index of the same table, do I have to use services such as DynamoDB and Redis? Thanks

```
### **Committing mechanisms for S3**

you should implement a context manager that acquires and releases a lock and then pass that to the `commit_lock` parameter of [`lance.write_dataset()`](https://lancedb.github.io/lance/api/python/lance.html#lance.write_dataset).

### **Concurrent Writer on S3 using DynamoDB**

By default, S3 does not support concurrent writes. Having two or more processes writing to the same table at the same time can lead to data corruption.

This is because S3, unlike other object stores, does not have any atomic put or copy operation.
```

**Comment by pavanramkumar** [comment]
This guide would be extremely valuable since I don't think it's widely documented elsewhere how to write a commit lock for s3. I'm long on the lance table format, but I'm struggling to implement multiple fan out steps in a data pipeline write to the same table safely.

**Comment by wjones127** [comment]
> since I don't think it's widely documented elsewhere how to write a commit lock for s3.

FYI, the commit lock will soon be unnecessary:

https://github.com/lancedb/lance/pull/3483

**Comment by pavanramkumar** [comment]
That's amazing! Looking forward to the next release. Any forthcoming documentation that shows how to do concurrent writes in python?

**Comment by wjones127** [comment]
> Any forthcoming documentation that shows how to do concurrent writes in python?

Do you mean distributed or concurrent? This ticket is about distributed writes, like something that would be done in a distributed engine like Spark. One client that is spreading out it's work amongst many machines.

Typically, when we talk about concurrent writes, we just mean many clients trying to make commits at once. I'm not sure there's much to document there. You just make write calls as usual from different computes.

The dynamodb / commit lock doesn't really have to do with distributed writes. The reason is usually distributes writes work by each worker staging data, and then only one commit happens at end of ETL job by a central node. So there aren't any concurrent commits happening to worry about.

**Comment by pavanramkumar** [comment]
Awesome, thanks for walking through that. My use case is certainly distributed writes (many fan out computes writing to the same table). 

If I understand right, #3483 now enables safe writes from multiple computes without the need for staging data and a central node to collect staged data and make a single commit.

Should I ignore this warning with `0.23.0` that accompanies `lance.write_dataset(...)`?

```
 WARN  lance_table::io::commit] Using unsafe commit handler. Concurrent writes may result in data loss. Consider providing a commit handler that prevents conflicting writes.
```

**Comment by wjones127** [comment]
> now enables safe writes from multiple computes without the need for staging data and a central node to collect staged data and make a single commit.

Yes, assuming that a partial change makes sense for your operation.

> Should I ignore this warning with 0.23.0 that accompanies lance.write_dataset(...)?

If you see that warning, you should pay attention to it. If you aren't concurrently writing then you should be okay. But in future versions that warning will go away and you won't have to worry about it.
